test/pricer/pricer.test.ts

```diff
@@ -0,0 +1,152 @@
+    let exchangeApiFactoryStub: SinonStub;
+
+    beforeEach(function () {
+      coinGeckoStub = createStubInstance(CoinGecko);
+      fakeRifUsd = randomBigNumberJs(25000);
+      fakeRbtcUsd = randomBigNumberJs(25000);
+      fakeRbtcRif = fakeRbtcUsd.dividedBy(fakeRifUsd);
+      fakeRifRbtc = fakeRifUsd.dividedBy(fakeRbtcUsd);
+      exchangeApiFactoryStub = stub(ExchangeApiFactory, 'getExchangeApi')
+        .withArgs('coinGecko')
+        .returns(coinGeckoStub);
+      stubExchangeRate(
+        coinGeckoStub,
+        RIF_SYMBOL,
+        INTERMEDIATE_CURRENCY,
+        fakeRifUsd
+      );
+      stubExchangeRate(
+        coinGeckoStub,
+        RBTC_SYMBOL,
+        INTERMEDIATE_CURRENCY,
+        fakeRbtcUsd
+      );
+    });
+
+    afterEach(function () {
+      restore();
+    });
+
+    function randomBigNumberJs(max: number) {
+      return BigNumberJs(Math.random() * max);
+    }
+
+    function stubExchangeRate(
+      exchangeApi: SinonStubbedInstance<BaseExchangeApi>,
+      sourceCurrency: string,
+      targetCurrency: string,
+      exchangeRate: BigNumberJs
+    ) {
+      exchangeApi.queryExchangeRate
+        .withArgs(sourceCurrency, targetCurrency)
+        .returns(Promise.resolve(exchangeRate));
+    }
+
+    it('should return exchange rate', async function () {
+      const exchangeRate = await getExchangeRate(RIF_SYMBOL, RBTC_SYMBOL);
+
+      expect(exchangeRate.toString()).to.be.equal(fakeRifRbtc.toString());
+    });
+
+    it('should return exchange rate if currencies are in lower case', async function () {
+      const rifSymbol = RIF_SYMBOL.toLowerCase();
+      const rbtcSymbol = RBTC_SYMBOL.toLowerCase();
+      stubExchangeRate(
+        coinGeckoStub,
+        rifSymbol,
+        INTERMEDIATE_CURRENCY,
+        fakeRifUsd
+      );
+      stubExchangeRate(
+        coinGeckoStub,
+        rbtcSymbol,
+        INTERMEDIATE_CURRENCY,
+        fakeRbtcUsd
+      );
+      const exchangeRate = await getExchangeRate(rifSymbol, rbtcSymbol);
+
+      expect(exchangeRate.toString()).to.be.equal(fakeRifRbtc.toString());
+    });
+
+    it('should return the reverse exchange rate', async function () {
+      const exchangeRate = await getExchangeRate(RBTC_SYMBOL, RIF_SYMBOL);
+
+      expect(exchangeRate.toString()).to.be.equal(fakeRbtcRif.toString());
+    });
+
+    it('should return exchange rate and update prioritization', async function () {
+      const coinCodexStub = createStubInstance(CoinCodex);
+      stubExchangeRate(
+        coinCodexStub,
+        RBTC_SYMBOL,
+        INTERMEDIATE_CURRENCY,
+        fakeRbtcUsd
+      );
+      exchangeApiFactoryStub.withArgs('coinCodex').returns(coinCodexStub);
+      coinGeckoStub.queryExchangeRate
+        .withArgs(RBTC_SYMBOL, INTERMEDIATE_CURRENCY)
+        .rejects();
+
+      const beforeTokenToApi = tokenToApi[RBTC_SYMBOL];
+      const exchangeRate = await getExchangeRate(RIF_SYMBOL, RBTC_SYMBOL);
+      const afterTokenToApi = tokenToApi[RBTC_SYMBOL];
```

Comment:
NIT: although it works, I think that we could have tested the same scenario by checking if the subsequent call is performed against the same pricer, instead of checking for an implementation detail

Reply:
cool, Ill do that. thanks

Reply:
updated